src/fileformats/ocd_georef_fields.cpp

```diff
@@ -0,0 +1,461 @@
+ * OcdCrsRecord struct is used for passing CRS information around within the
+ * translation code.
+ */
+struct OcdCrsRecord
+{
+	const OcdGrid ocd_grid_id; ///< OCD grid number
+	const int ocd_zone_id; ///< OCD zone number, may be negative
+};
+
+/**
+ * Provide possible OCD grid/zone combinations for a Mapper georeferencing.
+ *
+ * @see translateOcdToMapper
+ */
+std::vector<OcdCrsRecord> translateMapperToOcd(const MapperCrs crs_unique_id,
+                                               const int crs_param)
+{
+	std::vector<OcdCrsRecord> retval;
+
+	switch (crs_unique_id)
+	{
+	case MapperCrs::LOCAL:
+		retval.push_back({ OcdGrid::LOCAL, 0 });
+		break;
+	case MapperCrs::UTM:
+		if (std::abs(crs_param) >= 1 && std::abs(crs_param) <= 60)
+			retval.push_back({ OcdGrid::UTM, crs_param });
+		break;
+	case MapperCrs::GAUSS_KRUEGER:
+		if (crs_param >= 2 && crs_param <= 5)
+			retval.push_back({ OcdGrid::GERMANY, crs_param });
+		break;
+	case MapperCrs::EPSG:
+		// handle UTM separately, the rest is in the table
+		if (crs_param >= 32601 && crs_param <= 32660)
+			retval.push_back({ OcdGrid::UTM, crs_param - 32600 });
+		else if (crs_param >= 32701 && crs_param <= 32760)
+			retval.push_back({ OcdGrid::UTM, 32700 - crs_param });
+
+		for (const auto& record : grid_zone_epsg_codes)
+		{
+			if (record.epsg_code == crs_param)
+				retval.push_back({ record.ocd_grid_id, record.ocd_zone_id });
+		}
+		break;
+	}
+
+	return retval;
+}
+
+}  // anonymous namespace
+
+bool operator==(const OcdGeorefFields& lhs, const OcdGeorefFields& rhs)
+{
+	return lhs.i == rhs.i
+	        && lhs.m == rhs.m
+	        && lhs.x == rhs.x
+	        && lhs.y == rhs.y
+	        && ((qIsNaN(lhs.a) && qIsNaN(rhs.a)) || qAbs(lhs.a - rhs.a) < 5e-9) // 8-digit precision or both NaN's
+	        && lhs.r == rhs.r;
+}
+
+void OcdGeorefFields::setupGeoref(Georeferencing& georef,
+                                  const std::function<void (const QString&)>& warning_handler) const
+{
+	if (m >= 0)
+		georef.setScaleDenominator(m);
+
+	if (qIsFinite(a) && qAbs(a) >= 0.01)
+		georef.setGrivation(a);
+
+	if (r)
+		applyGridAndZone(georef, i, warning_handler);
+
+	QPointF proj_ref_point(x, y);
+	georef.setProjectedRefPoint(proj_ref_point, false);
+}
+
+OcdGeorefFields OcdGeorefFields::fromGeoref(const Georeferencing& georef,
+                                            const std::function<void (const QString&)>& warning_handler)
+{
+	OcdGeorefFields retval;
+
+	auto crs_id_string = georef.getProjectedCRSId();
+	auto crs_param_strings = georef.getProjectedCRSParameters();
+	auto report_warning = [&]() {
+		QStringList params;
+		std::for_each(begin(crs_param_strings), end(crs_param_strings),
+		              [&](const QString& s) { params.append(s); });
+		warning_handler(::OpenOrienteering::OcdFileExport::tr(
+		                    "Could not translate coordinate reference system '%1:%2'.")
+		                .arg(crs_id_string)
+		                .arg(params.join(QLatin1Char(','))));
```

Comment:
clazy suggests that `arg(crs_id_string, params.join(QLatin1Char(',')))` would be more efficient than calling `args` twice for arguments of type `QString`.

Reply:
One of 508 clazy warnings issued for this branch build. I'd say the benefits of optimizing an error path are shadowed by the effort required for the patch set rework. On the positive side, I tried out clazy and learned about existence of multi-arg `QString::arg()`.

Reply:
I run with `CLAZY_CHECKS` set to `level0,level1,no-rule-of-two-soft,no-const-signal-or-slot` in my Qt Creator build environment. I didn't try hard enough to put it in `code-check-wrapper.sh` or cmake settings.